node/src/common/AsyncQueue.h

```diff
@@ -0,0 +1,252 @@
+            if (!_isInitialized)
+            {
+                return false;
+            }
+
+            queueWasEmpty = _items.empty();
+            _items.push(std::forward<QueueItemType>(item));
+        }
+
+        // Only notify the worker thread if the queue is no longer empty
+        // (as that's the only way the worker thread would be waiting
+        // for this signal)
+        if (queueWasEmpty)
+        {
+            _actionRequired.notify_all();
+        }
+
+        return true;
+    }
+
+    // Waits until the queue is empty
+    void WaitForAll()
+    {
+        std::unique_lock<std::mutex> lock(_mutex);
+
+        _isEmpty.wait(lock, [this] {return (_items.empty() || _stopWorkerThread); });
+    }
+
+private:
+    class UnlockGuard
+    {
+    public:
+        UnlockGuard(std::unique_lock<std::mutex>& lock) : _lock(lock) { _lock.unlock(); }
+        ~UnlockGuard() { _lock.lock(); }
+
+    private:
+        UnlockGuard(const UnlockGuard&) = delete;
+        UnlockGuard& operator=(const UnlockGuard&) = delete;
+
+        std::unique_lock<std::mutex>& _lock;
+    };
+
+    void WaitForAndProcessItems()
+    {
+        std::unique_lock<std::mutex> lock(_mutex);
+
+        std::shared_ptr<IQueueItemHandler<QueueItem>> handler = _handler;
+        handler->OnStarted();
+
+        for (;;)
+        {
+            // Wait until either:
+            // - queue is not empty
+            // - worker thread has been asked to stop
+            _actionRequired.wait(lock, [this] { if (_items.empty()) _isEmpty.notify_all(); return (!_items.empty() || _stopWorkerThread); });
+            if (_stopWorkerThread)
+            {
+                break;
+            }
+
+            // Move the items to a temporary queue
+            std::queue<QueueItem> items(std::move(_items));
+
+            // Process all queue items outside the lock, the UnlockGuard will re-lock on destruction
+            UnlockGuard unlock(lock);
+
+            while (!items.empty())
+            {
+                // Pop the next item off and notify the registered worker to process it
+                QueueItem item = std::move(items.front());
+                items.pop();
+
+                try
+                {
+                    handler->OnProcessQueueItem(item);
+                }
+                catch (...)
+                {
+                    LogWarning("Caught exception while processing async queue items.");
+                }
+            }
+        }
+
+        try
+        {
+            handler->OnStopped();
+        }
+        catch (...)
+        {
+            LogWarning("Caught exception while stopping async queue handler.");
+        }
+
+        // Signal that the thread is done
+        _workerThreadStopped = true;
+        _actionRequired.notify_all();
+        //std::notify_all_at_thread_exit(_actionRequired, std::move(lock)); // This should replace the previous line but Android doesn't support it
```

Comment:
Seems like a good comment to keep in case someone tries to get clever without testing on Android.

---
In reply to: [71086966](https://github.com/openT2T/client/pull/1#discussion_r71086966) [](ancestors = 71086966)